TST, BUG: Re-raise MemoryError exception in test_large_zip's process

antlarr-suse opened the pull request:
Since #15893, test_large_zip's actual test is run in a child process,
so when this test raises a MemoryError exception, the exception is
lost and the @requires_memory decorator can't catch it to return
an xfail.

This commit uses a boolean variable in shared memory to flag if
the exception was raised, and in that case, re-raise it in the
parent process.

Fixes #16889

antlarr-suse committed:
Re-raise MemoryError exception in test_large_zip's process

Since #15893, test_large_zip's actual test is run in a child process,
so when this test raises a MemoryError exception, the exception is
lost and the @requires_memory decorator can't catch it to return
an xfail.

This commit uses a boolean variable in shared memory to flag if
the exception was raised, and in that case, re-raise it in the
parent process.

Fixes #16889

anirudh2290 commented on numpy/lib/tests/test_io.py:
```diff
@@ -574,16 +575,26 @@ def test_unicode_and_bytes_fmt(self, fmt, iotype):
     def test_large_zip(self):
-        def check_large_zip():
+        def check_large_zip(memoryerror_raised):
+            memoryerror_raised.value=False
```
nit: 
```suggestion
            memoryerror_raised.value = False
```

antlarr replied:
Oops, of course! I don't know how they slipped by. Fixed.

anirudh2290 commented on numpy/lib/tests/test_io.py:
```diff
@@ -574,16 +575,26 @@ def test_unicode_and_bytes_fmt(self, fmt, iotype):
+                try:
+                    np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
+                except MemoryError:
+                    memoryerror_raised.value=True
```
nit: 
```suggestion
                    memoryerror_raised.value = True
```

antlarr replied:
Fixed too

anirudh2290 reviewed (approved):
Added a few nits, overall LGTM ! Thank you for the fix @antlarr-suse !

antlarr-suse committed:
PEP-8 fixes

seberg commented:
Thanks! Was just going to merge it, but I think we may want to put the try/except clause to include the array creation, which can error out almost as plausibly?

antlarr-suse committed:
Catch MemoryError exceptions raised when allocating test_data

Move the try...except block out so it also catches the exceptions
raised when allocating test_data, as suggested by @seberg in #16890

antlarr commented:
> Thanks! Was just going to merge it, but I think we may want to put the try/except clause to include the array creation, which can error out almost as plausibly?

It never happened here, but indeed, in a system with less than ~2GBs available it can raise that exception too.

I've added that change and adjusted the lines so they are within 80 chars width

seberg commented:
Thanks @antlarr will putting it in.  Should be reasonable enough for the moment, although I suppose tweaking the decorator or skipping for 32bits could be an option on top of it.
